### DISCUSSION TITLE
Connection issues - Wireguard

### DISCUSSION THREAD

**Discussion by nunospot** [pivpn/pivpn#1766] [Support]
Hello, 

First of all, I've tried searching for an issue template, but I didn't find one. Sorry about that.

My problem is connecting to the internet through my pivpn setup. 

I have tried all the steps to "troubleshoot connection issues" from the FAQ section. It seems that my setup is not received any packages, I tested using tcpdump.

`pivpn -d` output
```
::: Generating Debug Output
::::            PiVPN debug              ::::
=============================================
::::            Latest commit            ::::
Branch: master
Commit: 16189edc7b03d3ed930dbb3cc908d8e2275a7563
Author: 4s3ti
Date: Thu Aug 3 23:33:23 2023 +0200
Summary: fix(core): typo on distroCheck
=============================================
::::        Installation settings        ::::
PLAT=Raspbian
OSCN=bullseye
USING_UFW=0
pivpnforceipv6route=1
IPv4dev=eth0
dhcpReserv=1
IPv4addr=192.168.50.152/24
IPv4gw=192.168.50.1
install_user=nuno-pi
install_home=/home/nuno-pi
VPN=wireguard
pivpnPORT=51820
pivpnDNS1=10.39.40.1
pivpnDNS2=
pivpnHOST=REDACTED
INPUT_CHAIN_EDITED=1
FORWARD_CHAIN_EDITED=0
INPUT_CHAIN_EDITEDv6=
FORWARD_CHAIN_EDITEDv6=
pivpnPROTO=udp
pivpnMTU=1420
pivpnDEV=wg0
pivpnNET=10.39.40.0
subnetClass=24
pivpnenableipv6=0
ALLOWED_IPS="0.0.0.0/0, ::0/0"
UNATTUPG=1
INSTALLED_PACKAGES=(grepcidr bsdmainutils iptables-persistent wireguard-tools qrencode unattended-upgrades)
=============================================
::::  Server configuration shown below   ::::
[Interface]
PrivateKey = server_priv
Address = 10.39.40.1/24
MTU = 1420
ListenPort = 51820
### begin client1 ###
[Peer]
PublicKey = client1_pub
PresharedKey = client1_psk
AllowedIPs = 10.39.40.2/32
### end client1 ###
### begin client2 ###
[Peer]
PublicKey = client2_pub
PresharedKey = client2_psk
AllowedIPs = 10.39.40.3/32
### end client2 ###
=============================================
::::  Client configuration shown below   ::::
[Interface]
PrivateKey = client1_priv
Address = 10.39.40.2/24
DNS = 10.39.40.1

[Peer]
PublicKey = server_pub
PresharedKey = client1_psk
Endpoint = REDACTED:51820
AllowedIPs = 0.0.0.0/0, ::0/0
=============================================
::::    Recursive list of files in       ::::
::::    /etc/wireguard shown below       ::::
/etc/wireguard:
configs
keys
wg0.conf

/etc/wireguard/configs:
client1.conf
client2.conf
clients.txt

/etc/wireguard/keys:
client1_priv
client1_psk
client1_pub
client2_priv
client2_psk
client2_pub
server_priv
server_pub
=============================================
::::            Self check               ::::
:: [OK] IP forwarding is enabled
:: [OK] Iptables MASQUERADE rule set
:: [OK] Iptables INPUT rule set
:: [OK] WireGuard is running
:: [OK] WireGuard is enabled
(it will automatically start on reboot)
:: [OK] WireGuard is listening on port 51820/udp
=============================================
:::: Having trouble connecting? Take a look at the FAQ:
:::: https://docs.pivpn.io/faq
=============================================
:::: WARNING: This script should have automatically masked sensitive       ::::
:::: information, however, still make sure that PrivateKey, PublicKey      ::::
:::: and PresharedKey are masked before reporting an issue. An example key ::::
:::: that you should NOT see in this log looks like this:                  ::::
:::: YIAoJVsdIeyvXfGGDDadHh6AxsMRymZTnnzZoAb9cxRe                          ::::
=============================================
::::            Debug complete           ::::
:::
::: Debug output completed above.
::: Copy saved to /tmp/debug.log
:::
```
`lsmod | grep wireguard` output
```
wireguard              77824  0
libchacha20poly1305    16384  1 wireguard
ip6_udp_tunnel         16384  1 wireguard
udp_tunnel             28672  1 wireguard
libcurve25519_generic    40960  1 wireguard
ipv6                  557056  56 wireguard
```

`cat /etc/pivpn/wireguard/setupVars.conf` output
```
PLAT=Raspbian
OSCN=bullseye
USING_UFW=0
pivpnforceipv6route=1
IPv4dev=eth0
dhcpReserv=1
IPv4addr=192.168.50.152/24
IPv4gw=192.168.50.1
install_user=nuno-pi
install_home=/home/nuno-pi
VPN=wireguard
pivpnPORT=51820
pivpnDNS1=10.39.40.1
pivpnDNS2=
pivpnHOST=XXX.XXX.XXX.XXX
INPUT_CHAIN_EDITED=1
FORWARD_CHAIN_EDITED=0
INPUT_CHAIN_EDITEDv6=
FORWARD_CHAIN_EDITEDv6=
pivpnPROTO=udp
pivpnMTU=1420
pivpnDEV=wg0
pivpnNET=10.39.40.0
subnetClass=24
pivpnenableipv6=0
ALLOWED_IPS="0.0.0.0/0, ::0/0"
UNATTUPG=1
INSTALLED_PACKAGES=(grepcidr bsdmainutils iptables-persistent wireguard-tools qrencode unattended-upgrades)
```
Note: Censuring my IP address

`ip -f inet address show eth0` output
```
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    inet 192.168.50.152/24 brd 192.168.50.255 scope global noprefixroute eth0
       valid_lft forever preferred_lft forever
```

Also check if the `curl -s https://checkip.amazonaws.com` matches my IP address, settled on `pivpnHOST`

`tcpdump -n -i eth0 udp port 51820` output

```
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
```

Answers to the last steps of the guide

- If you set up PiVPN with ethernet and later switched to wifi, you will have a different IP. Easiest way to fix is to reinstall and pick the new network interface.

No. It was always an ethernet connection.

- Check if your ISP uses Carrier-grade NAT (check online). With CGNAT, your router gets a private IP, making port forwarding ineffective. This is mostly the norm if your router connects via 4G/LTE. If that's the case, you need to ask the ISP for a public IP.

Not sure. I've searched the Internet and found no information to confirm or disconfirm that my ISP uses CGNAT. The public IP that I query at `https://checkip.amazonaws.com` is always the same. On CGNAT, does that change?

- If you see packets coming, but no response from the Pi, it may indicate routing issues, attempts to block the connection (on either side), or poor connectivity. In all cases, try to connect from a different network.

Not sure about this step. From the `tcpdump` output, it doesn't look like it is not receiving any packets, right?

- If you have multiple chained routers, then you need to configure multiple port forwardings. Example: (192.0.2.48) ISP router (192.168.1.1) ---> (192.168.1.2) Own router (192.168.23.1) ---> (192.168.23.211) Raspberry Pi. Given that, on the ISP router port forward 1194 udp to 192.168.1.2 and on your own router port forward 1194 UDP to 192.168.23.211

Yes. I've chained routers. This was my last attempt before opening an issue here, and it didn't do the trick either :(. 

From the ISP router
![image](https://github.com/pivpn/pivpn/assets/19820660/3ffd8608-7ef6-42e6-bd6a-d7e08f894331)

From my router
![image](https://github.com/pivpn/pivpn/assets/19820660/537e1b07-74c6-4f2e-9a02-953a8ad76a65)
![image](https://github.com/pivpn/pivpn/assets/19820660/49bd99d7-bf0d-4592-ab09-9cfecdd44256)

- You may have misconfigured firewall rules on your Pi, open an issue and add the output of sudo iptables -S and sudo iptables -t nat -S.

Done

I'm really not sure what I'm doing wrong :\

PS: Not sure if this is relevant or not, but I've Pi-hole installed, and I've also configured a DNS server on my router.

![image](https://github.com/pivpn/pivpn/assets/19820660/44d5523d-20e4-4947-9e50-53d92ee1b6ec)

If you need more information, please let me know.

**Comment by orazioedoardo** [comment]
The double port forwarding seems correct.

> Not sure. I've searched the Internet and found no information to confirm or disconfirm that my ISP uses CGNAT. The public IP that I query at https://checkip.amazonaws.com is always the same. On CGNAT, does that change?

On CGNAT, the IP from the site is not that of your NOS Net router but that of a router outside your network. Typically every router has the ability to see in the web UI its "internet address" or "wan address", in your language obviously. If the WAN address is the IP from the site then it's not CGNAT.

Another option, which I'm thinking of integrating into PiVPN, is running `traceroute -n ipfromthesite`. If you see line 1 with `192.168.50.1` and line 2 with `ipfromthesite`, then are probably not under CGNAT. If you see more lines between `192.168.50.1` and `ipfromthesite` you are probably under CGNAT.

**Comment by coolapso** [comment]
Pretty much what @orazioedoardo  said .... 
if I am not mistaken NOS uses CGNAT .... became a common practice in Portugal, last time I checked only Vodafone was not using it in certain areas ... but well things change. calling NOS they should be able to answer if you call them. 

one way you can ask them (in case they don't understand the CGNAT) is, if your public IP, the one websites and the internet sees is only yours or shared among multiple customers. 

When you are behind CGNAT your router sees a different WAN IP Than the websites see ...  What do you have on "Configuracao LAN" ??  I think there you should be able to track that too. 

feel free to hit me up on Matrix or IRC and I can help you in Portuguese if you need.

**Comment by nunospot** [comment]
Thank you for your fast replies! 

Don't know how but it's fixed :D Just tried to connect to the internet through my VPN and it's working.

Not sure if the CGNAT will be a problem in the future, let's see. For the time being, I will close this discussion.
